### ISSUE TITLE
Doubts in Demand Forecasting using TFT

### ISSUE THREAD

**Issue by hsugoken** [jdb78/pytorch-forecasting#521]
Hello
I am really thankful for the work that you have put in for pytorch-forecasting.
In the demand forecasting using TFT example:
1. `for idx in range(10):  # plot 10 examples`
   `best_tft.plot_prediction(new_x, new_raw_predictions, idx=idx, show_future_observed=False);`
I am not sure what these 10 examples are from, are they for different batch sizes?
2. In the Actuals vs Prediction:

> Checking how the model performs across different slices of the data allows us to detect weaknesses. Plotted below are the means of predictions vs actuals across each variable divided into 100 bins using the Now, we can directly predict on the generated data using the calculate_prediction_actual_by_variable() and plot_prediction_actual_by_variable() methods. The gray bars denote the frequency of the variable by bin, i.e. are a histogram.
 Does this mean to say that we are trying to tell how much these individual variables help in prediction? 

3.  `new_raw_predictions, new_x = best_tft.predict(new_prediction_data, mode="raw", return_x=True)`
What is difference between mode="raw" and mode="prediction">?

4. Also I checked the values in val_dataloader using `x,y in iter(dataloader)`. Here the `x['decoder_target']` has the target values that we want to predict already present. So:
       - in case of training this is being used for teacher forcing, right?
       - also when using predicting on val_dataloader will this part be still used or it will be ignored. (Is mode='prediction' during `.predict(...)` has some role to play in ensuring this doesn't gets used or `validation.to_dataloader(train=False, batch_size=batch_size * 10, num_workers=0)` does this?

Thank you.

**Comment by jdb78** [comment]
First of all: Great to hear you like the package! Let me try to answer your questions:

1.
> I am not sure what these 10 examples are from, are they for different batch sizes?
These are just the first 10 examples. If from the training dataset, they will be random selections.

2.
> Does this mean to say that we are trying to tell how much these individual variables help in prediction?
No, you this helps to understand in variable ranges, the model struggles to make good predictions. The importance of variables for prediction is another analysis.

3.
> What is difference between mode="raw" and mode="prediction">?
`"raw"` is about the raw output of the network which contains predictions, target scales, etc. while `"prediction"` is just prediction of the target values

4.
> Here the x['decoder_target'] has the target values that we want to predict already present. So:

It is actually not necessary for teacher-forcing but mostly for plotting.

**Comment by hsugoken** [comment]
Hi thank you for answering. I still have doubt with the 2nd question.

> No, you this helps to understand in variable ranges, the model struggles to make good predictions. The importance of variables for prediction is another analysis.

1. What do you mean be variable ranges? Are we plotting the histogram of variable values and corresponding to different bins we have overlayed the predictions there.
![image](https://user-images.githubusercontent.com/80521644/120833882-ee08ba80-c57f-11eb-85b3-fdd8f72f57f0.png)
 - In this plot, the gray shaded bars are the histogram for normalized discount in percentage?
 - How to infer from this if the model is being able to make a good prediction or not?

**Comment by jdb78** [comment]
1. By ranges, I mean that you can see in your examples that the model tends to not capture low predictions (what could be outliers). Generally, there is no range in the discount_in_percent variable that it seems to struggle with.
2. yes (right axis is frequency)
3. Looks like it is ok with the discount. If your model makes good predictions depends on the metrics and what you deem good. But there are also other considerations. I would check the dependence plots and also good and bad examples as in the tutorial to get an idea.

**Comment by hsugoken** [comment]
So in above diagram the tft is predicting for all encoder variables and then it is being plotted over the histogram of the same?

**Comment by jdb78** [comment]
1. It is for the variables in the decoder.
2. exactly

**Comment by hsugoken** [comment]
Thank you. So as the above plot is for the decoder block it helps us to understand which variables are useful for predicting based on the aforementioned plot.

**Comment by hsugoken** [comment]
Thank your for your patience. Also one last question regarding the plot above.
Here we are plotting the **discount_in_percent_averages** against the predictions. Here these predictions are for the **target variable** or **discount_in_percent_averages**?

**Comment by jdb78** [comment]
On the x-axis is the **discount_in_percent** and on the y-axis the average predicted and actual values for the **target variable**. Does this answer your question?

**Comment by hsugoken** [comment]
Yeah, thanks for clarifying all of my doubts. I was really stuck at what was happening here. Thank you! :)
